gadget/install/encrypt_test.go

```diff
@@ -58,78 +58,123 @@ func (s *encryptSuite) SetUpTest(c *C) {
-	c.Assert(err, IsNil)
-	c.Assert(dev.Node, Equals, "/dev/mapper/some-label")
-
-	c.Assert(s.mockCryptsetup.Calls(), DeepEquals, [][]string{
+func (s *encryptSuite) TestNewEncryptedDevice(c *C) {
+	for _, tc := range []struct {
+		formatErr error
+		openErr   string
+		err       string
+	}{
 		{
-			"cryptsetup", "-q", "luksFormat", "--type", "luks2", "--key-file", "-",
-			"--cipher", "aes-xts-plain64", "--key-size", "512", "--pbkdf", "argon2i",
-			"--iter-time", "1", "--label", "some-label-enc", "/dev/node1",
+			formatErr: nil,
+			openErr:   "",
+			err:       "",
 		},
 		{
-			"cryptsetup", "open", "--key-file", "-", "/dev/node1", "some-label",
+			formatErr: errors.New("format error"),
+			openErr:   "",
+			err:       "cannot format encrypted device: format error",
 		},
-	})
-
-	err = dev.Close()
-	c.Assert(err, IsNil)
-}
-
-func (s *encryptSuite) TestEncryptFormatError(c *C) {
-	s.mockCryptsetup = testutil.MockCommand(c, "cryptsetup", `[ "$2" == "luksFormat" ] && exit 127 || exit 0`)
-	s.AddCleanup(s.mockCryptsetup.Restore)
-
-	key := secboot.EncryptionKey{}
-	_, err := install.NewEncryptedDevice(&mockDeviceStructure, key, "some-label")
-	c.Assert(err, ErrorMatches, "cannot format encrypted device:.*")
-}
-
-func (s *encryptSuite) TestEncryptOpenError(c *C) {
-	s.mockCryptsetup = testutil.MockCommand(c, "cryptsetup", `[ "$1" == "open" ] && exit 127 || exit 0`)
-	s.AddCleanup(s.mockCryptsetup.Restore)
-
-	key := secboot.EncryptionKey{}
-	_, err := install.NewEncryptedDevice(&mockDeviceStructure, key, "some-label")
-	c.Assert(err, ErrorMatches, "cannot open encrypted device on /dev/node1:.*")
-}
-
-func (s *encryptSuite) TestEncryptAddKey(c *C) {
-	capturedFifo := filepath.Join(c.MkDir(), "captured-stdin")
-	s.mockCryptsetup = testutil.MockCommand(c, "cryptsetup", fmt.Sprintf(`[ "$1" == "luksAddKey" ] && cat %s/tmp-rkey > %s || exit 0`, dirs.SnapRunDir, capturedFifo))
-	s.AddCleanup(s.mockCryptsetup.Restore)
-
-	key := secboot.EncryptionKey{}
-	dev, err := install.NewEncryptedDevice(&mockDeviceStructure, key, "some-label")
-	c.Assert(err, IsNil)
-
-	rkey := secboot.RecoveryKey{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15}
-	err = dev.AddRecoveryKey(key, rkey)
-	c.Assert(err, IsNil)
-
-	c.Assert(s.mockCryptsetup.Calls(), DeepEquals, [][]string{
 		{
-			"cryptsetup", "-q", "luksFormat", "--type", "luks2", "--key-file", "-",
-			"--cipher", "aes-xts-plain64", "--key-size", "512", "--pbkdf", "argon2i",
-			"--iter-time", "1", "--label", "some-label-enc", "/dev/node1",
+			formatErr: nil,
+			openErr:   "open error",
+			err:       "cannot open encrypted device on /dev/node1: open error",
 		},
-		{
-			"cryptsetup", "open", "--key-file", "-", "/dev/node1", "some-label",
-		},
-		{
-			"cryptsetup", "luksAddKey", "/dev/node1", "-q", "--key-file", "-",
-			"--key-slot", "1", filepath.Join(dirs.SnapRunDir, "tmp-rkey"),
-		},
-	})
-	c.Assert(capturedFifo, testutil.FileEquals, rkey[:])
+	} {
```

Comment:
For completness we may want to have a test with both "openErr" and "formatErr" to have the full matrix

Reply:
`newEncryptedDevice` has a separate test which should ensure it's working properly, it shouldn't be necessary to test it again when testing the recovery key (`mockedOpenErr` and `mockedFormatErr` are relevant only when creating the device and not when adding a recovery key).